s3transfer/futures.py

```diff
@@ -285,44 +285,99 @@ def __init__(self, max_size, max_num_threads):
+        tag = kwargs.pop('future_tag', None)
+
         with self._lock:
+            # Wait if there is a maximum reached
+            self._wait_for_futures_to_complete_if_needed(tag)
+            # Submit the task to the executor
+            future = self._executor.submit(fn, *args, **kwargs)
+            # Add the recently added future to the currently running futures
+            self._add_future_to_currently_running(future, tag)
+        return future
+
+    def shutdown(self, wait=True):
+        self._executor.shutdown(wait)
+
+    def _get_tags_to_operate_on(self, tag):
+        # No matter the tag provided, each future is included under the 'all'
+        # tag to represent its membership in the set of all running futures.
+        tag_to_operate_on = [ALL_TAG]
+        # Only add the tag to the set of the tags to operate on if the tag
+        # is not equal to all and the tag is being tracked for max sizes.
+        if tag is not ALL_TAG and tag in self._tags_to_track:
+            tag_to_operate_on.append(tag)
+        return tag_to_operate_on
+
+    def _at_maximum_running_futures(self, tag):
+        return len(self._currently_running_futures[tag]) == \
+            self._max_sizes[tag]
+
+    def _wait_for_futures_to_complete_if_needed(self, tag):
+        tags_to_wait_on = self._get_tags_to_operate_on(tag)
+        for tag in tags_to_wait_on:
             # First determine if there are already too many futures running
-            if self._at_maximum_running_futures():
+            if self._at_maximum_running_futures(tag):
                 # If there are too many futures running, wait till some
                 # complete and save the remaining running futures as the
                 # next set to wait for.
-                self._currently_running_futures = futures.wait(
-                    self._currently_running_futures,
-                    return_when=futures.FIRST_COMPLETED
+                self._currently_running_futures[tag] = futures.wait(
+                        self._currently_running_futures[tag],
+                        return_when=futures.FIRST_COMPLETED
                 ).not_done
-            # Submit the task to the executor
-            future = self._executor.submit(fn, *args, **kwargs)
+
+    def _add_future_to_currently_running(self, future, tag):
+        tags_to_add_future_to = self._get_tags_to_operate_on(tag)
+        for tag in tags_to_add_future_to:
             # If the queue is bounded then add the new submitted task
             # to the list of futures to wait for. If it is not bounded,
             # then we do not need to store the future as we are not keeping
             # track of how many futures are currently being ran.
-            if self._max_size is not None:
-                self._currently_running_futures.add(future)
-        return future
+            if self._max_sizes[tag] != 0:
+                self._currently_running_futures[tag].add(future)
+
+
+class FutureTag(object):
```

Comment:
Did you intend for object identity to be a requirement for uniqueness?  Otherwise, a namedtuple would be good here.  You'd get equality and good reprs() out of the box.